mesonbuild/interpreter.py

```diff
@@ -3169,9 +3169,28 @@ def func_project(self, node, args, kwargs):
         if not self.is_subproject():
             self.build.project_name = proj_name
         self.active_projectname = proj_name
-        self.project_version = kwargs.get('version', 'undefined')
-        if not isinstance(self.project_version, str):
-            raise InvalidCode('The version keyword argument must be a string.')
+        version = kwargs.get('version', 'undefined')
+        if isinstance(version, list):
+            if len(version) != 1:
+                raise InvalidCode('Version argument is an array with more than one entry.')
+            version = version[0]
+        if isinstance(version, mesonlib.File):
+            self.add_build_def_file(version)
+            ifname = version.absolute_path(self.environment.source_dir,
+                                           self.environment.build_dir)
+            try:
+                ver_data = Path(ifname).read_text(encoding='utf-8').split('\n')
+            except FileNotFoundError:
+                raise InterpreterException('Version file not found.')
+            if len(ver_data) == 2 and ver_data[1] == '':
+                ver_data = ver_data[0:1]
```

Comment:
I would just `.rstrip('\n').split('\n')`, verify that it's exactly len() 1, and use that.

Reply:
That accepts files that have multiple newlines at the end, which arguably should be rejected.

Reply:
If your line ends with `\r\n` I think the `\r` would be included in the version string, no?

Reply:
```
>>> print('a\r\nb'.split('\n'))
['a\r', 'b']
>>> print('a\r\nb'.splitlines())
['a', 'b']
```

Reply:
read_text() uses universal_newlines

Reply:
oh didn't knowm good.

Reply:
https://docs.python.org/3/library/functions.html#open-newline-parameter